### PR TITLE
TechDocs: AWS S3 SDK version bump

### LINKED ISSUE #3944: TechDocs: AWS S3 SDK version bump
## Feature Suggestion

Currently the `aws-sdk` used in TechDocs is in version 2.
As v3 was just released it could be good to bump the sdk version.
[Link](https://aws.amazon.com/about-aws/whats-new/2020/12/aws-sdk-javascript-version-3-generally-available/)

Here are the benefits of v3:
- Add a better TypeScript support
- A modular architecture with a separate package for each service. 
It will allow us to do:
```ts
import { S3 } from ‘’@aws-sdk/client-s3;
```
 instead of:
```ts
import aws from 'aws-sdk';
```
- We no longer need to use `.promise()` which are not required

## Possible Implementation

Remove the aws sdk v2 and install the AWS service package for Amazon S3 `@aws-sdk/client-s3`.
Update`packages/techdocs-common/src/stages/publish/awsS3.ts` to use v3 and the S3 mock here `packages/techdocs-common/__mocks__/aws-sdk.ts` to match the new S3 signature.

## Context

PR which added Techdocs AWS S3 support #3794

**NOTE :** This feature is **already implemented**.
I could not include it in the PR because `aws-sdk-js-v3` had some TypeScript issues which were pending fix.
I will do a PR in the next few days.

### PR DESCRIPTION
**Pull request by ayshiff**
## Hey, I just made a Pull Request!

Closes #3944

This PR removes the aws sdk v2 and install the new AWS service package for Amazon S3 `@aws-sdk/client-s3`.

Updated files:
- `packages/techdocs-common/src/stages/publish/awsS3.ts`
- `packages/techdocs-common/__mocks__/aws-sdk.ts`

This PR also fix Windows tests that do not pass on master. 
Related PR #3925

I removed the `existsSync` check inside the publish mock implementation as the path that we provide to it is relative, and we can't retrieve the file from the mock.

```ts
const destination = `${entityRootDir}/${relativeFilePath}`; // S3 Bucket file relative path
```
Anyway, if the file does not exist, the `readFile` in the publish method will throw an error before.

Moreover, I also changed the `should fail to publish a directory` test as the path returned by the catch had a weird prefix `\\? \`, which is, I think, a `mock-fs` issue ([Link](https://github.com/backstage/backstage/pull/3925#issuecomment-755324921)).

Finally I noticed something a little weird in publishers.   
This behavior appears in the tests but I do not know if it is a behavior which could happen in another environment.

The `getFileTreeRecursively` method can return `\\` paths (on Windows) that looks like this : `\\directory\\relativeFilePath`.

The problem is that when we iterate on the file tree, we construct the filePath based on the fact that our path is in the Unix form `/directory/relativeFilePath` but that's not always the case.

The problematic line:
```ts
const relativeFilePath = filePath.replace(`${directory}/`, '');
```
Should we also add the replace statement for windows paths ?

```ts
const relativeFilePath = filePath.replace(`${directory}/`, '').replace(`${directory}\\`, '');
```

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: yarn.lock
Comment (OrkoHunter) [inline yarn.lock:25209]: About `yarn.lock` changes, I only notice additions. An entry for `aws-sdk` still exists. Perhaps if you run `yarn remove ...` to remove the old deps, it should update the `yarn.lock` as well.

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:183]: So much clean! 😁

File: yarn.lock
Comment (ayshiff) [reply yarn.lock]: After checking, it's normal that the dependency is still present since `catalog-backend` uses it in `plugins/catalog-backend/src/ingestion/processors/AwsOrganizationCloudAccountProcessor.ts`:

```ts
import AWS, { Organizations } from 'aws-sdk';
import { Account, ListAccountsResponse } from 'aws-sdk/clients/organizations';
```

Maybe we should update the sdk version (`@aws-sdk/client-organizations`) here too to stay consistent ?
Do you mind if I update it ?

Comment (OrkoHunter) [reply yarn.lock:25209]: For the version upgrade in catalog-backend, I think @backstage/maintainers can help us make the call!

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:51]: I tried to initialise without any `region` property and the aws client throws an error `Region is missing`

Looking at documentation, it feels that region is required for the aws-sdk v3. It will read the  `AWS_REGION` and `AWS_DEFAULT_REGION` environment variables if not provided here.

Just an FYI, I wanted to share. Will include this bit in https://github.com/backstage/backstage/issues/3946

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:143]: Hey! Unfortunately this might need something else here. Could you try opening some docs site locally? I am receiving `[object Object]` as response (for the Metadata API and for the request of `index.html` using `docsRouter()` below.

**Review by OrkoHunter — CHANGES_REQUESTED** [review]

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (ayshiff) [reply packages/techdocs-common/src/stages/publish/awsS3.ts:51]: Oh yes you are right !
I dit not pay enough attention to this change introduced in v3.

Do you want me to change it in this PR ?

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (ayshiff) [reply packages/techdocs-common/src/stages/publish/awsS3.ts:143]: I'll look at it more closely !

At the time of the implementation I had some questions regarding the `getObject` method https://github.com/backstage/backstage/pull/3794#discussion_r547181420.

I'm going to fix that and make some tests locally to make sure everything works 👍

Comment (OrkoHunter) [reply packages/techdocs-common/src/stages/publish/awsS3.ts:143]: Aah sorry, I missed that question in the previous PR!

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (OrkoHunter) [reply packages/techdocs-common/src/stages/publish/awsS3.ts:51]: I think it's fine to leave it as it is, and do it in "better AWS credentials handling" PR. :)

**Comment by ayshiff** [comment]
I redid tests locally and I fixed the way we got the data from `getObject`.
I added a `streamToString` helper which converts our readable stream (`file.Body`) to a string.

```ts
const streamToString = (stream: Readable): Promise<string> =>
  new Promise((resolve, reject) => {
    const chunks: any[] = [];
    stream.on('data', chunk => chunks.push(chunk));
    stream.on('error', reject);
    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
  });
```

The thing that I don't really like is that I have to cast the stream (which is of type `Readable | ReadableStream<any> | Blob`) to `Readable` to make TypeScript happy.

Moreover, I didn't add a check in case the Body is undefined but I don't really know in which case this error can appear.

**Comment by OrkoHunter** [comment]
Thank you for converting the stream. I can confirm that it works now.

> The thing that I don't really like is that I have to cast the stream (which is of type `Readable | ReadableStream<any> | Blob`) to `Readable` to make TypeScript happy.
> Moreover, I didn't add a check in case the Body is undefined but I don't really know in which case this error can appear.

Hmm. What bums me is that the official documentation does not have any code example for the simple use-case of reading a file. https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/s3-node-examples.html


IMO a `try/catch` on `streamToString` can be helpful here - 1. in case the response is not of type `Readable` and 2. when file.Body is missing.

**Review by OrkoHunter — APPROVED** [review]

File: packages/techdocs-common/__mocks__/@aws-sdk/client-s3.ts
Comment (freben) [inline packages/techdocs-common/__mocks__/@aws-sdk/client-s3.ts:64]: All of these could heavily benefit from `async`/`await`.

```ts
async putObject() {
  return '';
}
```

etc.

But not gonna block on that :) Just a readability thing

File: packages/techdocs-common/src/stages/publish/awsS3.test.ts
Comment (freben) [inline packages/techdocs-common/src/stages/publish/awsS3.test.ts:48]: just noting that it's not particularly safe to assume that these three pieces only contain filesystem safe characters :)

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (freben) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:28]: This can never throw. The Promise constructor is the only thing being invoked here. The actual function inside is run "inside" the promise, and will only lead to it being rejected. If I understand things correctly, you want this try/catch inside the promise instead.

File: packages/techdocs-common/src/stages/publish/awsS3.ts
Comment (freben) [inline packages/techdocs-common/src/stages/publish/awsS3.ts:158]: yeah this cast is unfortunate, but barring any better solution right now, let's continue and then tweak some other time

File: yarn.lock
Comment (freben) [inline yarn.lock:21313]: that's a crazy dep they have right there

**Review by freben — APPROVED** [review]
Some small comments, but slapping an approve on it to unblock

**Comment by freben** [comment]
Oh one more thing. This one does not have a changeset.

File: packages/techdocs-common/src/stages/publish/awsS3.test.ts
Comment (OrkoHunter) [reply packages/techdocs-common/src/stages/publish/awsS3.test.ts:48]: Curious what are the allowed characters for namespace, kind and name? TechDocs stores files in cloud storage and local, creating a nested directory like -> `namespace/kind/name/index.html`

Comment (freben) [reply packages/techdocs-common/src/stages/publish/awsS3.test.ts:48]: It's actually [configurable](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/service/CatalogBuilder.ts#L170).

But yeah we can just hope that nobody ever starts allowing slashes and colons and nulls and stuff inside their fields, because that is bound to break stuff all over the place.

Comment (ayshiff) [reply packages/techdocs-common/src/stages/publish/awsS3.test.ts:48]: Since you're talking about paths, do we fix the 
```ts
const relativeFilePath = filePath.replace(`${directory}/`, '');
``` 
line (present in the s3 and gcs publishers) I mentioned in the PR description above, which I think is a problem for Windows users.

Comment (freben) [reply packages/techdocs-common/src/stages/publish/awsS3.test.ts]: Hard to say without a windows machine in front of me, but it should probably be `path.relative(directory, filePath)` and handling leading slashes or dots accordingly.

Comment (ayshiff) [reply packages/techdocs-common/src/stages/publish/awsS3.test.ts]: As I think this problem is out of scope, I will create an issue to report the problem after re-testing on Windows.

File: yarn.lock
Comment (freben) [reply yarn.lock:25209]: Yeah @ayshiff if you want to contribute that in a separate PR we'd definitely appreciate it!

**Pull request merged**
